drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1226 @@
+{
+	modem_chat_release(&data->chat);
+
+	return modem_ppp_attach(data->ppp, data->dlci2_pipe);
+}
+
+static int modem_cellular_on_register_state_enter(struct modem_cellular_data *data)
+{
+	if (modem_chat_attach(&data->chat, data->dlci1_pipe) < 0) {
+		return -EAGAIN;
+	}
+
+	modem_cellular_start_timer(data, K_SECONDS(2));
+
+	return modem_chat_script_run(&data->chat, &net_stat_chat_script);
+}
+
+static bool modem_cellular_is_registered(struct modem_cellular_data *data)
+{
+	return (data->registration_status == 5) && (data->packet_service_attached == 1);
+}
+
+static void modem_cellular_register_event_handler(struct modem_cellular_data *data,
+						  enum modem_cellular_event evt)
+{
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_SUSPEND:
+		if (modem_cellular_gpio_is_enabled(&config->power_gpio)) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_POWER_OFF);
+		} else {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_IDLE);
+		}
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		if (modem_cellular_is_registered(data)) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_ROAMING);
+		}
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		modem_cellular_start_timer(data, K_SECONDS(2));
+
+		modem_chat_script_run(&data->chat, &net_stat_chat_script);
+
+		break;
+
+	default:
+		break;
+	}
+}
+
+static int modem_cellular_on_register_state_leave(struct modem_cellular_data *data)
+{
+	modem_cellular_stop_timer(data);
+
+	modem_chat_release(&data->chat);
+
+	return 0;
+}
+
+static int modem_cellular_on_roaming_state_enter(struct modem_cellular_data *data)
+{
+	net_if_up(modem_ppp_get_iface(data->ppp));
+
+	modem_chat_attach(&data->chat, data->dlci1_pipe);
+
+	modem_chat_script_run(&data->chat, &net_stat_chat_script);
+
+	modem_cellular_start_timer(data, K_SECONDS(4));
+
+	return 0;
+}
+
+static void modem_cellular_roaming_event_handler(struct modem_cellular_data *data,
+						 enum modem_cellular_event evt)
+{
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_SUSPEND:
+		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_CARRIER_OFF);
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		if (modem_cellular_is_registered(data) == false) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_RUN_DIAL_SCRIPT);
+		}
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		modem_chat_script_run(&data->chat, &net_stat_chat_script);
+
+		modem_cellular_start_timer(data, K_SECONDS(4));
+
+		break;
+
+	default:
+		break;
+	}
+}
+
+static int modem_cellular_on_roaming_state_leave(struct modem_cellular_data *data)
+{
+	modem_cellular_stop_timer(data);
+
+	modem_chat_script_abort(&data->chat);
+
+	modem_chat_release(&data->chat);
+
+	return 0;
+}
+
+static int modem_cellular_on_carrier_off_state_enter(struct modem_cellular_data *data)
+{
+	net_if_down(modem_ppp_get_iface(data->ppp));
```

Comment:
The driver should not really put the interface up/down. It should rather control the carrier/dormant state with `net_if_carrier/dormant_on/off()` (perhaps controlling dormant state is a better choice here, we use it for nRF91 modem, carrier is more of a physical carrier status).

Note, that interface won't be reported as ready to use (even if the application brings it up) unless you put it into carrier ON, dormant OFF.

Reply:
That would be the optimal solution, however I have not had much luck so far outside of this solution, so far trying to use net_ppp_carrier_on and net_ppp_carrier_off, which simply don't work. I will try to use net_if_dormant_on/off(), I have not tried those before :)

Reply:
Here is my progress towards fixing the PPP L2 subsystem https://github.com/zephyrproject-rtos/zephyr/pull/59124

Reply:
Updated as part of PPP L2 interface update